Source/WebSocket/WinHTTP/winhttp_websocket.cpp

```diff
@@ -0,0 +1,252 @@
+            }
+        }
+        catch (...)
+        {
+            hr = E_FAIL;
+            XAsyncComplete(message->asyncBlock, hr, sizeof(WebSocketCompletionResult));
+        }
+        return hr;
+    }
+
+    // Pull the next message from the queue and send it
+    HRESULT send_msg()
+    {
+        auto sendContext = http_allocate_shared<send_msg_context>();
+        sendContext->pThis = shared_from_this();
+        {
+            std::lock_guard<std::mutex> lock(m_outgoingMessageQueueLock);
+            ASSERT(!m_outgoingMessageQueue.empty());
+            sendContext->message = std::move(m_outgoingMessageQueue.front());
+            m_outgoingMessageQueue.pop();
+        }
+
+        auto hr = XAsyncBegin(sendContext->message.asyncBlock, shared_ptr_cache::store(sendContext), (void*)HCWebSocketSendMessageAsync, __FUNCTION__,
+            [](XAsyncOp op, const XAsyncProviderData* data)
+        {
+            WebSocketCompletionResult* result;
+            auto sendMsgContext = shared_ptr_cache::fetch<send_msg_context>(data->context, op == XAsyncOp::Cleanup);
+
+            switch (op)
+            {
+                case XAsyncOp::DoWork:
+                {
+                    return sendMsgContext->pThis->send_msg_do_work(&sendMsgContext->message);
+                }
+
+                case XAsyncOp::GetResult:
+                {
+                    result = reinterpret_cast<WebSocketCompletionResult*>(data->buffer);
+                    result->platformErrorCode = sendMsgContext->message.hr;
+                    result->errorCode = XAsyncGetStatus(data->async, false);
+                    return S_OK;
+                }
+
+                default: return S_OK;
+            }
+        });
+
+        if (SUCCEEDED(hr))
+        {
+            hr = XAsyncSchedule(sendContext->message.asyncBlock, 0);
+        }
+        return hr;
+    }
+
+
+    HRESULT disconnect_websocket(_In_ HCWebSocketCloseStatus closeStatus)
+    {
+        return m_httpTask->disconnect_websocket(closeStatus);
+    }
+
+    std::shared_ptr<xbox::httpclient::winhttp_http_task> m_httpTask;
+    HCCallHandle m_call = nullptr;
+
+    // Websocket state
+    std::mutex m_httpClientLock; // Guards access to HTTP socket
+    HCWebsocketHandle m_hcWebsocketHandle = nullptr;
+    std::mutex m_outgoingMessageQueueLock; // Guards access to m_outgoing_msg_queue
+    http_internal_queue<websocket_outgoing_message> m_outgoingMessageQueue; // Queue to order the sends   
+    std::atomic<int> m_numSends = 0; // Number of sends in progress and queued up.
+};
+
+
+HRESULT CALLBACK Internal_HCWebSocketConnectAsync(
+    _In_z_ const char* uri,
+    _In_z_ const char* subProtocol,
+    _In_ HCWebsocketHandle websocket,
+    _Inout_ XAsyncBlock* asyncBlock,
+    _In_ HCPerformEnv env
+    )
+{
+    assert(env != nullptr);
+
+    websocket->uri = uri;
+    websocket->subProtocol = subProtocol;
+    auto wsppSocket = http_allocate_shared<winhttp_websocket_impl>(websocket);
```

Comment:
seems like we should check that we aren't already connected here? Maybe set websocket->impl to null when we disconnect?  I don't think we are doing this in wspp but should probably be fixed in both places.

Reply:
You should be able to disconnect/reconnect without tossing handle so impl != null isn't right logic, but I'll add a TODO to guard against double connect